firebase-firestore/src/main/java/com/google/firebase/firestore/local/QueryEngine.java

```diff
@@ -87,7 +97,39 @@ public ImmutableSortedMap<DocumentKey, Document> getDocumentsMatchingQuery(
       return result;
     }
 
-    return executeFullCollectionScan(query);
+    QueryContext context = new QueryContext();
+    result = executeFullCollectionScan(query, context);
+    if (result != null && automaticIndexingEnabled) {
+      CreateCacheIndexes(query, context, result.size());
+    }
+    return result;
+  }
+
+  /**
+   * Decides whether SDK should create a full matched field index for this query based on query
+   * context and query result size.
+   */
+  // TODO(csi): Auto experiment data.
+  private void CreateCacheIndexes(Query query, QueryContext context, int resultSize) {
+    String decisionStr = "";
+    // If evaluation is updated, please update tests in SQLiteLocalStoreTest.java
+    if (context.getDocumentReadCount() > 2 * resultSize) {
```

Comment:
I really think we should hardcode some absolute minimum document read count that must be exceeded before creating an index. What do you think?

Reply:
Could you be more specific what `minimum document read count` refers to?

If it is minimum document reads in full collection scan. I believe there is beneficial of reading 1 result from 3 documents collection. And this difference can be solved by heuristic.

if it is minimum result. Is it also useful when result size is 0. If we using index, then we don't need to go through all the documents to find out there is nothing matched.

Reply:
Discussed in the meeting, we are going to set the threshold to 100.